server/readonly/readonly_service.ts

```diff
@@ -0,0 +1,105 @@
+/*
+ *   Copyright OpenSearch Contributors
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License").
+ *   You may not use this file except in compliance with the License.
+ *   A copy of the License is located at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   or in the "license" file accompanying this file. This file is distributed
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+import { URL } from 'url';
+import {
+  Logger,
+  OpenSearchDashboardsRequest,
+  SessionStorageFactory,
+} from '../../../../src/core/server';
+import { globalTenantName, isPrivateTenant } from '../../common';
+import { SecurityClient } from '../backend/opensearch_security_client';
+import { IAuthenticationType, OpenSearchAuthInfo } from '../auth/types/authentication_type';
+import { SecuritySessionCookie } from '../session/security_cookie';
+import { SecurityPluginConfigType } from '../index';
+import { ReadonlyService as BaseReadonlyService } from '../../../../src/core/server/security/readonly_service';
+
+export class ReadonlyService extends BaseReadonlyService {
+  private readonly logger: Logger;
+  private readonly securityClient: SecurityClient;
+  private readonly auth: IAuthenticationType;
+  private readonly securitySessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>;
+  private readonly config: SecurityPluginConfigType;
+
+  constructor(
+    logger: Logger,
+    securityClient: SecurityClient,
+    auth: IAuthenticationType,
+    securitySessionStorageFactory: SessionStorageFactory<SecuritySessionCookie>,
+    config: SecurityPluginConfigType
+  ) {
+    super();
+    this.logger = logger;
+    this.securityClient = securityClient;
+    this.auth = auth;
+    this.securitySessionStorageFactory = securitySessionStorageFactory;
+    this.config = config;
+  }
+
+  isAnonymousPage(request: OpenSearchDashboardsRequest) {
+    if (!request.headers || !request.headers.referer) {
+      return false;
+    }
+
+    try {
+      const url = new URL(request.headers.referer as string);
+      const pathsToIgnore = ['login', 'logout', 'customerror'];
+      return pathsToIgnore.includes(url.pathname?.split('/').pop() || '');
+    } catch (error: any) {
+      this.logger.error(`Could not parse the referer for the capabilites: ${error.stack}`);
+    }
+  }
+
+  isReadOnlyTenant(authInfo: OpenSearchAuthInfo): boolean {
+    const currentTenant = authInfo.user_requested_tenant || globalTenantName;
+
+    // private tenant is not affected
+    if (isPrivateTenant(currentTenant)) {
+      return false;
+    }
+
+    return authInfo.tenants[currentTenant] !== true;
```

Comment:
Lets use a variable to make it clear what is being extracted from the auth info data structure
```suggestion
    let readWriteAccess = authInfo.tenants[currentTenant];
    return !readWriteAccess;
```